flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedHaServicesWithLeadershipControl.java

```diff
@@ -36,16 +40,23 @@ public EmbeddedHaServicesWithLeadershipControl(Executor executor) {
         this(
                 executor,
                 new PerJobCheckpointRecoveryFactory<EmbeddedCompletedCheckpointStore>(
-                        (maxCheckpoints, previous) -> {
+                        (maxCheckpoints, previous, stateRegistryFactory, ioExecutor) -> {
+                            List<CompletedCheckpoint> checkpoints =
+                                    previous != null
+                                            ? previous.getAllCheckpoints()
+                                            : Collections.emptyList();
+                            SharedStateRegistry stateRegistry =
```

Comment:
This is something the original PR tried to change. The original PRs intention was not to recreate the `SharedStateRegistry` if we are not recreating the `CompletedCheckpointStore`.

Reply:
`EmbeddedCompletedCheckpointStore` **is** created a few lines below:
```
                            SharedStateRegistry stateRegistry =
                                    stateRegistryFactory.create(ioExecutor, checkpoints);
                            if (previous != null) {
                                if (!previous.getShutdownStatus().isPresent()) {
                                    throw new IllegalStateException(
                                            "Completed checkpoint store from previous run has not yet shutdown.");
                                }
                                return new EmbeddedCompletedCheckpointStore(
                                        maxCheckpoints, checkpoints, stateRegistry);
                            }
                            return new EmbeddedCompletedCheckpointStore(
                                    maxCheckpoints, checkpoints, stateRegistry);
```
Or do you mean something else?

Reply:
Just in terms of classes. In fact it is not recreated. IMO, we could even simply return `previous`. Not sure why we actually don't do it. 

The intention was not to recreate the `SharedStateRegistry` if we have `previous`, because the one in previous should already be populated with all of the completed checkpoints. Populating the `SharedStateRegistry` turns out to be an expensive operation that the user would prefer to omit.

Reply:
> Just in terms of classes. In fact it is not recreated

Sorry I don't get you: `EmbeddedCompletedCheckpointStore` object is recreated with `new` and it represents the store, doesn't it?
The checkpoints from the old store **are** reused, but they are immutable; the registry, OTH could be closed, so I think it would be error-prone to reuse it.

I think the intention is to not re-create the registry in `CheckpointCoordinator.restoreLatestCheckpointedStateInternal`;
(currently, it re-uses the store but re-creates the registry).

Btw, I see that `EmbeddedHaServicesWithLeadershipControl` is used only in tests.

Reply:
Yes, you are obviously right. I was talking bollocks. I got completely confused :( 

Sorry for the confusion.

Reply:
No worries, thanks for raising these concerns :slightly_smiling_face: